code/data_structures/src/tree/binary_tree/avl_tree/avl__tree.py

```diff
@@ -0,0 +1,262 @@
+        A = self.node 
+        B = self.node.left.node 
+        T = B.right.node 
+        
+        self.node = B 
+        B.right.node = A 
+        A.left.node = T 
+
+    
+    def lrotate(self):
+        # Rotate left pivoting on self
+        debug ('Rotating ' + str(self.node.key) + ' left') 
+        A = self.node 
+        B = self.node.right.node 
+        T = B.left.node 
+        
+        self.node = B 
+        B.left.node = A 
+        A.right.node = T 
+        
+            
+    def update_heights(self, recurse=True):
+        if not self.node == None: 
+            if recurse: 
+                if self.node.left != None: 
+                    self.node.left.update_heights()
+                if self.node.right != None:
+                    self.node.right.update_heights()
+            
+            self.height = max(self.node.left.height,
+                              self.node.right.height) + 1 
+        else: 
+            self.height = -1 
+            
+    def update_balances(self, recurse=True):
+        if not self.node == None: 
+            if recurse: 
+                if self.node.left != None: 
+                    self.node.left.update_balances()
+                if self.node.right != None:
+                    self.node.right.update_balances()
+
+            self.balance = self.node.left.height - self.node.right.height 
+        else: 
+            self.balance = 0 
+
+    def delete(self, key):
+        # debug("Trying to delete at node: " + str(self.node.key))
+        if self.node != None: 
+            if self.node.key == key: 
+                debug("Deleting ... " + str(key))  
+                if self.node.left.node == None and self.node.right.node == None:
+                    self.node = None # leaves can be killed at will 
+                # if only one subtree, take that 
+                elif self.node.left.node == None: 
+                    self.node = self.node.right.node
+                elif self.node.right.node == None: 
+                    self.node = self.node.left.node
+                
+                # worst-case: both children present. Find logical successor
+                else:  
+                    replacement = self.logical_successor(self.node)
+                    if replacement != None: # sanity check 
+                        debug("Found replacement for " + str(key) + " -> " + str(replacement.key))  
+                        self.node.key = replacement.key 
+                        
+                        # replaced. Now delete the key from right child 
+                        self.node.right.delete(replacement.key)
+                    
+                self.rebalance()
+                return  
+            elif key < self.node.key: 
+                self.node.left.delete(key)  
+            elif key > self.node.key: 
+                self.node.right.delete(key)
+                        
+            self.rebalance()
+        else: 
+            return 
```

Comment:
? What's the point of this?

Reply:
For rebalancing tree

Reply:
No I meant why the else clause in the first place? An empty return does nothing.